browser/brave_ads/creatives/search_result_ad/creative_search_result_ad_tab_helper.cc

```diff
@@ -0,0 +1,255 @@
+
+  if (!web_contents->GetBrowserContext() ||
+      web_contents->GetBrowserContext()->IsOffTheRecord()) {
+    return;
+  }
+
+  if (!ShouldSupportSearchResultAds()) {
+    return;
+  }
+
+  CreateForWebContents(web_contents);
+}
+
+bool CreativeSearchResultAdTabHelper::ShouldHandleCreativeAdEvents() const {
+  if (!ShouldSupportSearchResultAds()) {
+    // If the feature is disabled, we should not trigger ad events.
+    return false;
+  }
+
+  if (ShouldAlwaysTriggerSearchResultAdEvents()) {
+    // If the feature is enabled, we should always trigger ad events.
+    return true;
+  }
+
+  // If the feature is enabled, we should only trigger ad events when the user
+  // has joined Brave Rewards.
+  const Profile* const profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  return profile->GetPrefs()->GetBoolean(brave_rewards::prefs::kEnabled);
+}
+
+void CreativeSearchResultAdTabHelper::MaybeHandleCreativeAdClickedEvent(
+    const GURL& url) {
+  CHECK(ShouldHandleCreativeAdEvents());
+
+  if (creative_search_result_ad_handler_) {
+    creative_search_result_ad_handler_->MaybeTriggerCreativeAdClickedEvent(url);
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+AdsService* CreativeSearchResultAdTabHelper::GetAdsService() const {
+  if (g_ads_service_for_testing) {
+    CHECK_IS_TEST();
+
+    return g_ads_service_for_testing;
+  }
+
+  Profile* const profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  return AdsServiceFactory::GetForProfile(profile);
+}
+
+void CreativeSearchResultAdTabHelper::
+    MaybeExtractCreativeAdPlacementIdsAndHandleViewedEvents() {
+  if (!ShouldHandleCreativeAdEvents()) {
+    return;
+  }
+
+  if (!creative_search_result_ad_handler_) {
+    return;
+  }
+
+  creative_search_result_ad_handler_->ExtractCreativeAdPlacementIdsFromWebPage(
+      web_contents()->GetPrimaryMainFrame(),
+      base::BindOnce(
+          &CreativeSearchResultAdTabHelper::MaybeHandleCreativeAdViewedEvents,
+          weak_factory_.GetWeakPtr()));
+}
+
+void CreativeSearchResultAdTabHelper::MaybeHandleCreativeAdViewedEvents(
+    std::vector<std::string> placement_ids) {
+  for (const auto& placement_id : placement_ids) {
+    MaybeHandleCreativeAdViewedEvent(placement_id);
+  }
+}
+
+void CreativeSearchResultAdTabHelper::MaybeHandleCreativeAdViewedEvent(
+    const std::string& placement_id) {
+  CHECK(!placement_id.empty());
+
+  const std::string javascript = base::ReplaceStringPlaceholders(
+      kDataPlacementIdVisibilityCheckJavaScript, {placement_id}, nullptr);
+
+  web_contents()->GetPrimaryMainFrame()->ExecuteJavaScriptInIsolatedWorld(
+      base::UTF8ToUTF16(javascript),
+      base::BindOnce(&CreativeSearchResultAdTabHelper::
+                         MaybeHandleCreativeAdViewedEventCallback,
+                     weak_factory_.GetWeakPtr(), placement_id),
+      ISOLATED_WORLD_ID_BRAVE_INTERNAL);
```

Comment:
@aseren is working on https://github.com/brave/brave-browser/issues/39404 which will address concerns from @diracdeltas. This PR simply cleans up the code.